cub/test/catch2_test_device_reduce_nondeterministic.cu

```diff
@@ -0,0 +1,298 @@
+
+  /// Cache load modifier for reading input elements
+  static constexpr cub::CacheLoadModifier LOAD_MODIFIER = cub::CacheLoadModifier::LOAD_DEFAULT;
+  constexpr static int ITEMS_PER_THREAD                 = NOMINAL_ITEMS_PER_THREAD_4B;
+  constexpr static int BLOCK_THREADS                    = NOMINAL_BLOCK_THREADS_4B;
+};
+
+template <int ItemsPerThread, int BlockSize>
+struct hub_t
+{
+  struct Policy : cub::ChainedPolicy<300, Policy, Policy>
+  {
+    constexpr static int ITEMS_PER_THREAD = ItemsPerThread;
+
+    using ReducePolicy = AgentReducePolicy<BlockSize, ItemsPerThread>;
+
+    // SingleTilePolicy
+    using SingleTilePolicy = ReducePolicy;
+
+    // SegmentedReducePolicy
+    using SegmentedReducePolicy = ReducePolicy;
+
+    // ReduceNondeterministicPolicy
+    using ReduceNondeterministicPolicy = ReducePolicy;
+  };
+
+  using MaxPolicy = Policy;
+};
+
+C2H_TEST("Nondeterministic Device reduce works with float and double on gpu",
+         "[reduce][nondeterministic]",
+         float_type_list)
+{
+  using type          = typename c2h::get<0, TestType>;
+  const int num_items = 1 << 20;
+  c2h::device_vector<type> d_input(num_items);
+  c2h::gen(C2H_SEED(2), d_input, static_cast<type>(-1000.0), static_cast<type>(1000.0));
+
+  c2h::device_vector<type> d_output(1);
+
+  const auto env = cuda::execution::require(cuda::execution::determinism::not_guaranteed);
+  cub::DeviceReduce::Reduce(d_input.begin(), d_output.begin(), num_items, cuda::std::plus<type>{}, type{}, env);
+
+  c2h::host_vector<type> h_input = d_input;
+
+  c2h::host_vector<type> h_expected(1);
+  h_expected[0] = std::accumulate(h_input.begin(), h_input.end(), type{}, ::cuda::std::plus<type>());
+
+  REQUIRE_APPROX_EQ_EPSILON(h_expected, d_output, type{0.01});
+}
+
+C2H_TEST("Nondeterministic Device reduce works with float and double on gpu with NaN",
+         "[reduce][nondeterministic]",
+         float_type_list)
+{
+  using type     = typename c2h::get<0, TestType>;
+  using limits_t = cuda::std::numeric_limits<type>;
+
+  constexpr int num_items = 1 << 20;
+
+  constexpr auto min_op = cuda::minimum<type>{};
+  constexpr auto init   = cuda::std::numeric_limits<type>::max();
+
+  c2h::device_vector<type> d_input(num_items);
+  c2h::gen(C2H_SEED(2), d_input, static_cast<type>(-1000.0), static_cast<type>(1000.0));
+
+  const int num_indices = GENERATE_COPY(take(2, random(1 << 10, num_items)));
+
+  CAPTURE(num_indices, num_items);
+
+  // generate random indices to scatter NaNs
+  c2h::device_vector<int> indices(num_indices);
+  for (int i = 0; i < 2; ++i)
+  {
+    const type nan_val = i == 0 ? limits_t::signaling_NaN() : limits_t::quiet_NaN();
+
+    c2h::gen(C2H_SEED(2), indices, 0, num_items - 1);
+    auto begin = thrust::make_constant_iterator(nan_val);
+    auto end   = begin + num_indices;
+
+    // sprinkle some NaNs randomly throughout the input
+    thrust::scatter(c2h::device_policy, begin, end, indices.cbegin(), d_input.begin());
```

Comment:
Important: I think this causes a data race, since `indices` may contain duplicated indices leading to concurrent writes. If you need unique indices, consider generating a sequence and shuffling it.
```c++
thrust::sequence(indices.begin(), indices.end());
thurst::shuffle(indices.begin(), indices.end(), rng);
```